### DISCUSSION TITLE
/api/users and speaker sort

### DISCUSSION THREAD

**Discussion by glasnt** [pretalx/pretalx#1157] [Q&A] [answered]
More of a series of questions than a bug. Not critical to solve, more of a question of how it all works together.

As an admin with permission to do all the things, on `/orga/event/EVENT/submissions/SUBID/speakers/` I get the option to add speakers by searching for them. 

If I search for a substring, I can see a network call to `/orga/event/EVENT/api/users?search=TERM`. **Where is that code?** I know the frontend uses `typeahead`, but I'm wondering where `/api/users` is defined. 

As when I search for myself, I see multiples of myself: 

<img width="524" alt="A conference of katies" src="https://user-images.githubusercontent.com/813732/131490671-1538fe8a-b602-4123-ab4b-a298eb34bbf8.png">

I haven't submitted a talk for this event, and I don't come up in the speakers. Indeed one of those is a reviewer email that I only use for anonymous reviewing validation. But also, I have multiple of the same listing. This view shows 5 entries, but the API results for this search is `count: 10`, also with more value duplication (guessing there's a limit to the number of results returned). 

I also ask as this API returns in `results` a list of of email and name dicts. How are those associated to `/orga/event/EVENT/speakers` profiles? By email? 

Those entities use numeric identifiers (rather than 6-digit codes externally). For sessions with multiple speakers, they are sorted on /speakers and the pretalx web schedule talk page by numeric speaker, but on the schedule page they're not sorted the same.  (Happy to provide a data sample for this one, but I don't want to share it on a public form)

**Comment by rixx** [comment, marked as answer]
> More of a series of questions than a bug.

I'm going to disagree with you on that 😬

An answer in two parts:

**How it works**: The API is supposed to search for speakers who have an account with *any event you have access to*, but accidentally also caught regular organiser accounts for those events (because the same view was also meant to be used to add users to organiser teams. Which didn't work.). Fix is under way!

**How it fits together**: The user search API is currently not meant to be used by external consumers, especially as we're currently in the process of switching to a different prediction library (you can already see this in the new email sending form!). It uses email addresses as identifier rather than speaker codes because it is meant to work on a cross-event basis – and the six-letter identifiers are event-scoped.

*Edit:* I meant to link to the code via commit message, but apparently commits don't get linked on GitHub discussions. The API code is [here](https://github.com/pretalx/pretalx/blob/master/src/pretalx/orga/views/person.py)
